connector/connect/common/src/main/protobuf/spark/connect/commands.proto

```diff
@@ -421,6 +424,32 @@ message StreamingQueryManagerCommandResult {
   }
 }
 
+// The protocol for client-side StreamingQueryListener.
+// This command will only be set when either the first listener is added to the client, or the last
+// listener is removed from the client.
+// The add_listener_bus_listener command will only be set true in the first case.
+// The remove_listener_bus_listener command will only be set true in the second case.
+message StreamingQueryListenerBusCommand {
+  oneof command {
+    bool add_listener_bus_listener = 1;
+    bool remove_listener_bus_listener = 2;
+  }
+}
+
+// The protocol for the returned events in the long-running response channel.
+// The event_type for QueryProgressEvent is 1;
+// for QueryTerminatedEvent is 2; for QueryIdleEvent is 3;
+// There is no QueryStartedEvent defined here,
+// it is added as a field in WriteStreamOperationStartResult
+message StreamingQueryListenerEvents {
+  string event_json = 1;
```

Comment:
Is this guaranteed to be JSON?

Reply:
Yes all streaming query events have a json method https://github.com/apache/spark/blob/fd518dac8d9f9a117e06319e768c2c80f8ce600f/sql/core/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryListener.scala#L127

on client there are fromJson methods to reconstruct the event:
https://github.com/apache/spark/blob/bbf06c867e0e2f68d3b3b86080e54a16e6213235/python/pyspark/sql/streaming/listener.py#L194